File: app/src/main/java/com/rockspin/bargainbits/ui/activities/WatchListPresenter.java
package com.rockspin.bargainbits.ui.activities;

import com.rockspin.bargainbits.ui.BasePresenter;
import javax.inject.Inject;
import rx.Observable;

public class WatchListPresenter extends BasePresenter<WatchListPresenter.View, Void, Void> {

    @Inject public WatchListPresenter() { /* not used */ }

    @Override public void start(View view) {
        super.start(view);

        addSubscription(view.onListLoading().subscribe(view::showListLoading));
        addSubscription(view.onBackButtonPressed().subscribe(aVoid -> view.closeWatchList()));
    }

    public interface View {
        Observable<Boolean> onListLoading();

        Observable<Void> onBackButtonPressed();

        void closeWatchList();

        void showListLoading(boolean loading);
    }
}

File: app/src/main/kotlin/com/rockspin/bargainbits/ui/watch_list/WatchListPresenter.kt
package com.rockspin.bargainbits.ui.watch_list

import com.rockspin.bargainbits.ui.mvp.BaseMvpPresenter
import javax.inject.Inject


class WatchListPresenter @Inject constructor(): BaseMvpPresenter<WatchListView>() {

    override fun onViewCreated(view: WatchListView) {
        super.onViewCreated(view)

        addLifetimeDisposable(view.onListIsLoading().subscribe({
            loading ->
            run {
                view.showLoadingView(loading)

                if (loading) {
                    view.showListEmptyView(false)
                } else {
                    if (view.isDealsEmpty()) {
                        view.showListEmptyView(true)
                    } else {
                        view.showListEmptyView(false)
                    }
                }
            }
        }))
        addLifetimeDisposable(view.onBackButtonPressed().subscribe({ view.closeWatchList() }))
    }
}